src/transport_udp.cpp

```diff
@@ -1,237 +1,515 @@
+    }
+
+    std::lock_guard<std::mutex> _(_connections_mutex);
+
+    auto conn_it = conn_contexts.find(conn_id);
+    if (conn_it != conn_contexts.end()) {
+        send_disconnect(conn_it->second->id, conn_it->second->addr);
 
-  // Copy the address information
-  std::memcpy(addr, &it->second.addr, sizeof(it->second.addr));
+        delegate.on_connection_status(conn_it->second->id, TransportStatus::Disconnected);
 
-  return true;
+        addr_conn_contexts.erase(conn_it->second->addr.id);
+        conn_contexts.erase(conn_it);
+
+        delegate.on_connection_status(last_conn_id, TransportStatus::Ready);
+    }
+
+    if (!isServerMode) { // Client mode, stop threads
+        stop = true;
+    }
 }
 
-void
-UDPTransport::close(const TransportConnId& context_id)
-{
+AddrId UDPTransport::create_addr_id(const sockaddr_storage &addr) {
+    AddrId id;
+
+    switch (addr.ss_family) {
+        case AF_INET: {
+            sockaddr_in *s = (sockaddr_in *) &addr;
+
+            id.port = s->sin_port;
+            id.ip_lo = s->sin_addr.s_addr;
+            break;
+        }
+        default: {
+            // IPv6
+            sockaddr_in6 *s = (sockaddr_in6 *) &addr;
 
-  if (isServerMode) {
-    addrKey ak;
+            id.port = s->sin6_port;
 
-    addr_to_key(remote_contexts[context_id].addr, ak);
+            id.ip_hi = (uint64_t) &s->sin6_addr;
+            id.ip_lo = (uint64_t) &s->sin6_addr + 8;
+            break;
+        }
+    }
 
-    remote_addrs.erase(ak);
-    remote_contexts.erase(context_id);
-    dequeue_data_map.erase(context_id);
-  }
+    return std::move(id);
 }
 
-void
-UDPTransport::addr_to_key(sockaddr_storage& addr, addrKey& key)
-{
+TransportRemote UDPTransport::create_addr_remote(const sockaddr_storage &addr) {
+    TransportRemote remote;
+
+    char ip[INET6_ADDRSTRLEN];
+
+    remote.proto = TransportProtocol::UDP;
 
-  key.port = 0;
-  key.ip_lo = 0;
-  key.ip_hi = 0;
+    switch (addr.ss_family) {
+        case AF_INET: {
+            sockaddr_in *s = (sockaddr_in *) &addr;
 
-  switch (addr.ss_family) {
-    case AF_INET: {
-      sockaddr_in* s = (sockaddr_in*)&addr;
+            remote.port = s->sin_port;
+            inet_ntop(AF_INET, &s->sin_addr, ip, sizeof(ip));
+            break;
+        }
+        default: {
+            // IPv6
+            sockaddr_in6 *s = (sockaddr_in6 *) &addr;
 
-      key.port = s->sin_port;
-      key.ip_lo = s->sin_addr.s_addr;
-      break;
+            remote.port = s->sin6_port;
+            inet_ntop(AF_INET6, &s->sin6_addr, ip, sizeof(ip));
+            break;
+        }
     }
-    default: {
-      // IPv6
-      sockaddr_in6* s = (sockaddr_in6*)&addr;
 
-      key.port = s->sin6_port;
+    return std::move(remote);
+}
+
+bool UDPTransport::send_connect(const TransportConnId conn_id, const Addr& addr) {
+    UdpProtocol::ConnectMsg chdr {};
+
+    chdr.idle_timeout = 120;
+
+    std::lock_guard<std::mutex> _(_socket_write_mutex);
+
+    int numSent = sendto(fd,
```

Comment:
It's to sync both fd_reader and fd_writer threads on writes and updating the shared Connection context.  Not all the methods pass in the connection context right now, but they do pass in the addr.  they main thing that needs to be in sync is the connection context, so some of them can be removed but it's nice to know that sending protocol messages that both fd_writer and fd_reader will do will be in sync on order and when they are writing/updating data.

Reply:
I'm missing the point, I guess. I see no reason to have the mutex locked while calling an OS-level `sendto()` call.

Reply:
You are missing the point.  As mentioned previously it's not about the `sendto()`. The `send_*` methods update the connection context. While a few do not pass in the connection context, the others do.

Maybe your comment is that the ones that do not update it, do not need it... while that is true and it can be removed, it's there right now because it will likely be added back. Regardless, as I mentioned above, I'll remove these in favor of the connection mutex.

Reply:
What I don't like about moving locking out of the `send_*` methods is that they become non thread-safe.  It will require the caller to lock or be super careful. I'll update the method comments to mention that.

Reply:
In the case, what's to protect? I don't see any reason for the lock.

More generally, I would avoid locking while calling into the OS to send data. Whatever time that takes, the lock is held with no benefit.  I get it that sometimes that's unavoidable.

On locking in every function, while you're right that can be a problem, it can be documented. You don't want the cost of locking and unlocking all the time.